src/corehost/common/utils.cpp

```diff
@@ -463,3 +463,26 @@ pal::string_t get_dotnet_root_from_fxr_path(const pal::string_t &fxr_path)
     pal::string_t fxr_root = get_directory(fxr_dir);
     return get_directory(get_directory(fxr_root));
 }
+
+#define TEST_ONLY_MARKER "d38cc827-e34f-4453-9df4-1e796e9f1d07"
+
+// Retrieves environment variable which is only used for testing.
+// This will only work if there's also a _dotnet_test_only_enabled file next to the current module
+// (which means next to dotnet.exe, apphost.exe, hostfxr.dll and so on, whichever is being tested)
+bool test_only_getenv(const pal::char_t* name, pal::string_t* recv)
+{
+    // This is a static variable which is embeded in the product binary (somewhere).
+    // The marker values is a GUID so that it's unique and can be found by doing a simple search on the file
+    // The first character is used as the decider:
+    //  - Default value is 'd' (stands for disabled) - test only behavior is disabled
+    //  - To enable test-only behaviors set it to 'e' (stands for enabled)
+    constexpr size_t EMBED_SIZE = sizeof(TEST_ONLY_MARKER) / sizeof(TEST_ONLY_MARKER[0]);
+    volatile static char embed[EMBED_SIZE] = TEST_ONLY_MARKER;
+
+    if (embed[0] != 'e')
+    {
+        return false;
+    }
```

Comment:
I am not convinced that a smart compiler couldn't remove this code during optimization.

Why is this any better than a simple `#ifdef TEST_ONLY` ?  Either way we need to alter the binary...

Reply:
The `embed` is marked as `volatile` so compiler is not allowed to optimize the memory load away. In fact originally I forgot about it and clang did optimize it away.

The difference from `#ifdef` is that this allows us to test the binaries produced by official build which will be used by everything else. This saves us:
* Trouble with having to produce two sets of build outputs
* Potential differences between the two builds - which might cause product issues

I understand that we're not testing exactly (byte by byte) the same bits since we're patching the binary, but
* The patching is tiny (and we're not changing code itself)
* We only do this in relatively small number of tests, vast majority of our tests are using the actually official binaries (byte-by-byte)